kb/development/developer-builds.md

```diff
@@ -0,0 +1,727 @@
+
+```
+make -C ../source1/build-foo/Makefile SOURCE_DIR=$PWD dep
+```
+
+#### Refresh an existing build directory after changing dependencies
+
+Suppose that you have a source tree with an existing build tree, and you modify some C source files in a way that affects the build: new source files, removed source files, or changed include directives.
+
+CMake generally picks up updated dependencies automatically. (Are there limitations?) So just rebuild as usual.
+
+With mbedtls-prepare-build, dependencies are analyzed only when you run the `mbedtls-prepare-build` command. If dependencies have changed, run `make dep` (or its alias `make prepare`) in the build directory.
+
+```
+mbedtls-prepare-build -d build-foo
+make -C build-foo
+# ... hackity hack ...
+# E.g. add a new source file
+make -C build-foo dep
+make -C build-foo
+```
+
+If some files have been removed, you should run `make clean` in the build directory before re-running `make` or `make dep`. Otherwise obsolete build products may remain in the build tree, and they may confuse some scripts that enumerate build products with wildcards.
+
+#### Refresh an existing build directory after changing generated files
+
+CMake automatically rebuilds configuration-independent intermediate files as needed.
+
+Limitation: mbedtls-prepare-build does not understand how the configuration-independent intermediate files are built. If they have changed, you need to regenerate them manually in the source tree first. (On first use, mbedtls-prepare-build generates all the configuration-independent intermediate files, but it doesn't detect when this needs to be re-done.)
+
+In Mbed TLS, to update any intermediate file that needs updating:
+
+```
+make generated_files
+```
+
+In TF-PSA-Crypto, or in Mbed TLS, to unconditionally rebuild all intermediate files:
+
+```
+framework/scripts/make_generated_files.py
+```
+
+If build dependencies have changed in the generated files, [run `make dep` in the build tree](#refresh-an-existing-build-directory-after-changing-dependencies).
+
+#### Refresh an existing build directory after changing branches
+
+After changing branches, or rebasing an existing branch, the build scripts may have changed.
+
+With CMake, ???
+
+With mbedtls-prepare-build, run `make dep`  (or its alias `make prepare`) in the build directory in case the available files have changed. In some cases you should run `make clean` first, to avoid leaving obsolete build products behind. Furthermore, [you may need to refresh the generated files in the source tree](refresh-an-existing-build-directory-after-changing-generated-files).
+
+In Mbed TLS, to update any intermediate file that needs updating:
+
+```
+make generated_files
+make -C build-foo clean dep
+```
+
+In TF-PSA-Crypto, or in Mbed TLS, to unconditionally rebuild all intermediate files:
+
+```
+framework/scripts/make_generated_files.py
+make -C build-foo clean dep
+```
+
+### Builds with drivers
+
+#### Build with all drivers
+
+CMake: ???
```

Comment:
This is a known missing feature which we'll resolve eventually for the sake of `all.sh`.